Objective-C/CBLDatabase.mm

```diff
@@ -332,7 +332,23 @@ - (BOOL) inBatch: (NSError**)outError usingBlock: (void (NS_NOESCAPE ^)())block
         if (!transaction.begin())
             return convertError(transaction.error(), outError);
         
-        block();
+        @try {
+            NSError* err = nil;
+            block(&err);
+            if (err) {
+                // if swift throws an exception, `err` will be populated
+                transaction.abort();
+                return createError(CBLErrorUnexpectedError,
+                                   [NSString stringWithFormat: @"%@", err.localizedDescription],
+                                   outError);
+            }
+        } @catch(NSException* e) {
+            // if objc block throws an exception, control reaches here
+            transaction.abort();
+            return createError(CBLErrorUnexpectedError,
+                               [NSString stringWithFormat: @"%@ %@ %@", e.name, e.reason, e.userInfo],
+                               outError);
+        }
```

Comment:
My point is... this part of the code shouldn't exist.... particularly because of "if objc block throws an exception, control reaches here". Converting an exception (particularly one originating from code you don't control) is a crash waiting to happen down the line.

The Swift part of this is fine.

Reply:
> Converting an exception (particularly one originating from code you don't control) is a crash waiting to happen down the line

If any NSExceptions are raised in client code(inside in..batch block in objc), which will get caught in this catch block rather than crash(see Objective-C/Tests/DatabaseTest.m), cancels the transaction and returns unexpected error.

Reply:
I believe what @dtorres is saying is you're very broadly swallowing _all_ exceptions here even though some of them should raised as a valid exception that stops program execution in order to prevent corruption (or very difficult to debug crashes) further down in the stack.

To reiterate his comment above:
>Exceptions in Obj-C are complicated but the gist is, you shouldn't catch them in a generic context, but only in a case by case basis where you know it's ok as most NSExceptions are thrown due to unrecoverable errors (out of memory, memory corruption, internal inconsistency, etc) 

Are there specific exceptions you're expecting here that are safe to catch versus others that should be raised and cause program execution to stop?

Reply:
See examples here of rethrowing if the local context doesn't understand / need to process what it's caught: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Exceptions/Tasks/NestingExceptionHandlers.html#//apple_ref/doc/uid/20000060-CJBBFDIF

And the Clang article you linked above mentions this as well:

>The standard Cocoa convention is that exceptions signal programmer error and are not intended to be recovered from. Making code exceptions-safe by default would impose severe runtime and code size penalties on code that typically does not actually care about exceptions safety.

as well as memory leaks actually being fine in this context where needed
>**Therefore, ARC-generated code leaks by default on exceptions, which is just fine if the process is going to be immediately terminated anyway. Programs which do care about recovering from exceptions should enable the option.**

Reply:
Lastly, from an old Apple dev: https://stackoverflow.com/a/3379240/308315

>Any exception that passes through system framework code will leave said framework in an undefined state.. Catching said exceptions and trying to recover from it will lead to memory leaks, undefined behaviour and crashing.

>Just because the language has syntax to support something, doesn't mean the frameworks are designed to support that thing, too. The frameworks have a very firm policy: exceptions are unrecoverable errors. Trying to catch and recover goes against framework design and, quite clearly, an exception that "jumps over" framework based stack frames is a violation of documented policy and will yield unspecified behavior.

Reply:
got it now, thanks for pointing it out and detailed explanation. 

https://github.com/couchbase/couchbase-lite-ios/commit/ec919108a9f38deb2684f7cb9d3f8596787ac140